Add simulation of concurrent make appointments

OlehKozlovskyi opened the pull request.

OlehKozlovskyi committed:
Implement asyncronius reading file

OlehKozlovskyi committed:
Try implemet async writing

OlehKozlovskyi committed:
Implement asyncronius reading

OlehKozlovskyi committed:
Remove unused file

OlehKozlovskyi committed:
Implement DoctorRepository

OlehKozlovskyi committed:
Implememnt booking hour by patient, also add indexer to doctor

OlehKozlovskyi committed:
Modify main menu

OlehKozlovskyi committed:
Implement DoctorService

OlehKozlovskyi committed:
Make IClinicRepository methods asyncronic

OlehKozlovskyi committed:
make DoctorRepositoruAsync

OlehKozlovskyi committed:
Implement ClinicService

OlehKozlovskyi committed:
Finish modify app architecture

OlehKozlovskyi committed:
Refactor class data

OlehKozlovskyi committed:
Refactor DocotrRepository and DoctorService

OlehKozlovskyi committed:
Modify ClinicService and Repository(move null checkers and lockers on application layer)

OlehKozlovskyi committed:
Implement Logger

OlehKozlovskyi committed:
Implement Concurrent access simulation

OlehKozlovskyi committed:
Fix bug with concurrent make appointments

OlehKozlovskyi committed:
Rename namespace of domain entities

sawabond commented on NetworkOfPrivateClinics/NetworkOfPrivateClinics/BusinessLogic/ClinicRepository.cs:
```diff
@@ -6,27 +6,36 @@
         {
-            _context = clinicContext;
+            await Task.Run(() =>
+            {
+                _clinics.Add(clinic);
+            });
         }
 
-        public void DeleteClinic(int id)
+        public async Task DeleteClinicAsync(Clinic clinic)
         {
-            Clinic clinic = _context.First(x => x.ClinicID == id);
-            _context.Remove(clinic);
+            await Task.Run(() =>
+            {
+                _clinics.Remove(clinic);
+            });
         }
```
No needed. Why would u make it async if u work with in memory storage. It's not I/O bound operation

sawabond commented on NetworkOfPrivateClinics/NetworkOfPrivateClinics/BusinessLogic/ClinicRepository.cs:
```diff
@@ -6,27 +6,36 @@
-
-        public IEnumerable<Clinic> GetClinics() => _context;
+        public async Task<Clinic> GetClinicByIdAsync(int id)
+        {
+            return await Task.FromResult(_clinics.First(x => x.ClinicID == id));
+        }
 
-        public void InsertClinic(Clinic clinic) => _context.Add(clinic);
+        public async Task<List<Clinic>> GetClinicsAsync()
+        {
+            return await Task.FromResult(_clinics);
+        }
```
No need to make async

sawabond commented on NetworkOfPrivateClinics/NetworkOfPrivateClinics/BusinessLogic/Doctor.cs:
```diff
@@ -36,5 +38,26 @@ public Doctor(int id, string name, string surname, DoctorType type, decimal cost
+
+        public async Task<bool> TryMakeAppointmentAsync(int dayNumber, TimeOnly hour, Patient patient)
+        {
+            var appointmentBook = false;
```
```suggestion
            var appointmentBooked = false;
```

sawabond commented on NetworkOfPrivateClinics/NetworkOfPrivateClinics/BusinessLogic/Doctor.cs:
```diff
@@ -36,5 +38,26 @@ public Doctor(int id, string name, string surname, DoctorType type, decimal cost
+            }
+        }
+
+        public async Task<bool> TryMakeAppointmentAsync(int dayNumber, TimeOnly hour, Patient patient)
```
No need to make async

sawabond commented on NetworkOfPrivateClinics/NetworkOfPrivateClinics/BusinessLogic/DoctorRepository.cs:
```diff
@@ -0,0 +1,74 @@
+        private readonly List<Doctor> _doctors = new();
+        private readonly ICustomLogger _logger;
+
+        public DoctorRepository(ICustomLogger logger)
```
Looks like u don't need to make any async method here

sawabond commented on NetworkOfPrivateClinics/NetworkOfPrivateClinics/Program.cs:
```diff
@@ -101,14 +105,40 @@ private void MainMenuManageOptions(int optionsNumber)
+        {
+            ProjectLogger<Program> logger = new();
+            DoctorService doctorService = new(new DoctorRepository(logger));
+            ConcarentAccessSimulation simulation = new(doctorService);
```
```suggestion
            ConcurrentAccessSimulation simulation = new(doctorService);
```

sawabond commented on NetworkOfPrivateClinics/NetworkOfPrivateClinics/Program.cs:
```diff
@@ -101,14 +105,40 @@ private void MainMenuManageOptions(int optionsNumber)
+            await Task.WhenAll(tasks);
+        }
+
+        private async Task<(Doctor, Patient)> GetIncomingDataForSimulation()
+        {
+            Doctor doctor = new DoctorsFactory(1234, "Alice", "Johnson", DoctorType.Neurologist, 100m,
+                        new AppointmentsFactory("8:00", "17:00")).GetDoctor();
+            Patient patient = new(100, "Oleh", "Kozlovjiy", "Oleg@ukr.net", "+380 68 857 7128");
+            return await Task.FromResult((doctor, patient));
+        }
```
No need to be async

sawabond reviewed (changes requested):
Requested changes

OlehKozlovskyi committed:
Convert async methods

Convert async several asyncronous methods to syncronous in Doctor, DoctorRepository, DoctorService classes

OlehKozlovskyi committed:
Convert async clinic methods

Convert async methods on syncronous in ClinicService, ClinicRepository classes

OlehKozlovskyi committed:
Modify concurrent access simulation and logger

sawabond reviewed (approved):
Good job, correct
